examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+    return 0;
+  }
+
+  // Note that for the gradient we need to return a Tensor<1,dim>
+  template <>
+  Tensor<1,2>
+  Solution<2>::gradient (const Point<2> &p,
+                         const unsigned int component) const
+  {
+    using numbers::PI;
+    double x = p(0);
+    double y = p(1);
+    Tensor<1,2> return_value;
+    if (component == 0)
+      {
+        return_value[0] = PI * cos (PI * x);
+        return_value[1] = 0.0;
+      }
+    else if (component == 1)
+      {
+        return_value[0] = y * PI * PI * sin( PI * x);
+        return_value[1] = - PI * cos (PI * x);
+      }
+    else if (component == 2)
+      {
+        return_value[0] = PI * cos (PI * x) * cos (PI * y);
+        return_value[1] =  - PI * sin (PI * x) * sin(PI * y);
+      }
+    else
+      Assert (false, ExcMessage ("Component out of range in Solution"));
+    return return_value;
+  }
+
+  template <>
+  Tensor<1,3>
+  Solution<3>::gradient (const Point<3> &p,
+                         const unsigned int component) const
+  {
+    using numbers::PI;
+    double x = p(0);
+    double y = p(1);
+    double z = p(2);
+    Tensor<1,3> return_value;
+    if (component == 0)
+      {
+        return_value[0] = 2 * PI * cos (PI * x);
+        return_value[1] = 0.0;
+        return_value[2] = 0.0;
+      }
+    else if (component == 1)
+      {
+        return_value[0] = y * PI * PI * sin( PI * x);
+        return_value[1] = - PI * cos (PI * x);
+        return_value[2] = 0.0;
+      }
+    else if (component == 2)
+      {
+        return_value[0] = z * PI * PI * sin( PI * x);
+        return_value[1] = 0.0;
+        return_value[2] = - PI * cos (PI * x);
+      }
+    else if (component == 3)
+      {
+        return_value[0] = PI * cos (PI * x) * cos (PI * y) * sin (PI * z);
+        return_value[1] =  - PI * sin (PI * x) * sin(PI * y) * sin (PI * z);
+        return_value[2] = PI * sin (PI * x) * cos (PI * y) * cos (PI * z);
+      }
+    else
+      Assert (false, ExcMessage ("Component out of range in Solution"));
+    return return_value;
+  }
+
+  // Implementation of $f$. See the introduction for more information.
+  template <int dim>
+  class RightHandSide : public Function<dim>
+  {
+  public:
+    RightHandSide () : Function<dim>(dim+1) {}
+
+    virtual double value (const Point<dim> &p,
+                          const unsigned int component = 0) const;
+  };
+
+  template <>
+  double
+  RightHandSide<2>::value (const Point<2> &p,
+                           const unsigned int component) const
+  {
+    using numbers::PI;
+    double x = p(0);
+    double y = p(1);
+    if (component == 0)
+      return PI * PI * sin(PI * x) + PI * cos(PI * x) * cos(PI * y);
+    if (component == 1)
+      return - PI * PI * PI * y * cos(PI * x) - PI * sin(PI * y) * sin(PI * x);
+    if (component == 2)
+      return 0;
+
+    Assert (false, ExcMessage ("Component out of range in RightHandSide"));
```

Comment:
If you move the assertion to the top, it will already crash there. The `return 0;` at the end of the function is then, strictly speaking, dead code that can never be reached, but it is necessary to avoid compiler warnings about non-`void` functions that may not reach a `return` statement (because the compiler cannot see that every other case is already covered).

I've sometimes done things like
```
  return std::numeric_limits<double>::NaN();
```
(or whatever the exact spelling is) to really make sure that the value that is returned is completely unusable. It certainly also serves a documentation purpose in that case.